### PR TITLE
Add pprof labels in processes and for lifecycles

### PR DESCRIPTION
**Pull request by zeripath**
Use pprof labelling to help identify goroutines with processes.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/graceful/manager.go [hunk 1]
```diff
@@ -157,7 +132,7 @@ func (g *Manager) RunAtTerminate(terminate func()) {
 }
 
 // RunAtShutdown creates a go-routine to run the provided function at shutdown
-func (g *Manager) RunAtShutdown(ctx context.Context, shutdown func()) {
+func (g *Manager) RunAtShutdown(done context.Context, shutdown func()) {
```
Comment (wxiaoguang) [inline modules/graceful/manager.go]: I think we should keep the name `ctx`. and `case <-ctx.Done()` seems better than `case <-done.Done()`

File: modules/graceful/manager_windows.go [hunk 2]
```diff
@@ -73,6 +74,16 @@ func (g *Manager) start() {
 	g.hammerCtx, g.hammerCtxCancel = context.WithCancel(g.ctx)
 	g.doneCtx, g.doneCtxCancel = context.WithCancel(g.ctx)
 
+	// Next add pprof labels to these contexts
+	g.terminateCtx = pprof.WithLabels(g.terminateCtx, pprof.Labels("graceful-lifecycle", "with-terminate"))
+	g.shutdownCtx = pprof.WithLabels(g.shutdownCtx, pprof.Labels("graceful-lifecycle", "with-shutdown"))
+	g.hammerCtx = pprof.WithLabels(g.hammerCtx, pprof.Labels("graceful-lifecycle", "with-hammer"))
+	g.doneCtx = pprof.WithLabels(g.doneCtx, pprof.Labels("graceful-lifecycle", "with-manager"))
```
Comment (wxiaoguang) [inline modules/graceful/manager_windows.go]: doneCtx -> managerCtx ?

**Review by wxiaoguang — APPROVED** [review]
Generally LGTM

**Review by KN4CK3R — APPROVED** [review]

**Comment by zeripath** [comment]
make lgtm work

**Comment by wxiaoguang** [comment]
The outputed label looks like this, what's the meaning of these `-2`, `-3` ... suffix?

image

**Comment by zeripath** [comment]
Look at the way PIDs are generated.

We should probably switch to using unixtimenano instead as the base.

**Pull request merged**